torch_xla/csrc/xla_sharding_util.cpp

```diff
@@ -382,4 +383,52 @@ std::vector<at::Tensor> ShardingUtil::ShardTensor(
   return shards;
 }
 
+void ShardingUtil::PrepareOutputShardingPropagation(
+    std::vector<XLATensorPtr>* tensors, absl::Span<const size_t> indices,
+    ComputationPtr computation,
+    std::vector<torch::lazy::BackendDataPtr>* data_placeholders,
+    std::vector<XLATensor::ShardingSpecPtr>* sharding_specs) {
+  auto computation_proto = computation->computation().proto();
+  auto tuple_shardings =
+      computation_proto.spmd_output_sharding().tuple_shardings();
+  if (tuple_shardings.empty()) {
+    return;
+  }
+  XLA_CHECK(indices.size() == tuple_shardings.size())
+      << "Expected " << indices.size() << " sharding annotations, but got "
+      << tuple_shardings.size();
+
+  for (int i = 0; i < indices.size(); ++i) {
+    auto xtensor = (*tensors)[indices[i]];
+    if (tuple_shardings[i].type()) {
+      (*sharding_specs)[i] = std::make_shared<XLATensor::ShardingSpec>(
+          tuple_shardings[i],
+          MakeShapeWithDeviceLayout(
+              xtensor->shape().get(),
+              static_cast<XlaDeviceType>(xtensor->GetDevice().type())));
+      xtensor->SetShardingSpec(*(*sharding_specs)[i]);
+
+      // Create sharded data placeholder, this will be used to
+      // hold the corresponding computation results.
+      auto sharded_data_placeholder =
```

Comment:
In `xla_graph_executor.cpp`, the sharded data preparation for implicitly replicated data is removed. Should we still prepare a sharded placeholder regardless of `tuple_shardings[i].type()`?

Reply:
In summary, we shouldn't prepare the sharded data placeholder in the graph executor, but do so using this `ShardingUtil` API. This happens based on `tuple_shardings[i].type`, not "regardless of" it. Um, let me know if you still have questions.

Reply:
Just for my understanding - in the implicit replication case, the data is wrapped in `PjRtShardedData` and `xtensor->sharding_spec()` remains nullptr. In the `OutputHandler`, we're wrapping the resulting shards in `PjRtShardedData`. If we don't override the placeholder with `PjRtShardedData` here, we'll be trying to `Assign` a `PjRtShardedData` to a `PjRtData` placeholder.

If this is working, it seems like the `tuple_shardings[i].type()` is non-false even for the implicitly replicated case, unless I'm missing something?

Reply:
After merging with the recent implicit replication changes, all the outputs in case of sharded computation should be wrapped inside PjRtShardedData. This is a bug, will update the PR.